doc/manual/contract.md

```diff
@@ -0,0 +1,329 @@
+
+The implementation of contracts is recommended when you are dealing with business rules. For example when you are developing a library it may be necessary to specify contracts to guarantee valid use of the API. Also, contracts support inheritance, which guarantees that contracts will be satisfied in future specification.
+
+Let us take a look at the `Stack` class bellow:
+
+~~~
+	class Stack[T]
+		invariant(size >= 0)
+
+		var data = new Array[T]
+
+		var size = 0
+
+		fun is_empty: Bool do return size == 0
+
+		# Add the given argument `object` on top of the stack
+		fun push(object: T)
+		is
+			ensure(size == old(size) + 1)
+			ensure(data[size-1] == object)
+		do
+			top += 1
+			data[top] = object
+		end
+
+		# Remove the last item on the stack
+		fun pop
+		is
+			expect(not is_empty)
+			ensure(size == old(size) - 1)
+		do
+			data.pop
+			top -= 1
+		end
+
+		# Return the element at the given index.
+		fun get_at(index: Int): T
+		is
+			expect(index < size)
+			ensure(result == self[index])
+		do
+			return self[index]
+		end
+	end
+~~~
+
+As you can see the contracts are defined using several annotations. `invariant` which designates the class invariants, `expect` which designates the preconditions and `ensure` which designates the postconditions. The conditions of a contract must be expressed using boolean expressions in the same way as assertions. Each contract of this example will be detailed in a section below.
+
+All Contracts conditions can be declared in several ways, all are equivalent:
+~~~
+	fun foo(a: Int, b: Int) is ensure(a > 0, b < 0)
+
+	fun foo is ensure(a > 0 and b < 0)
+
+	fun foo is ensure(a > 0), ensure(b < 0)
+~~~
+
+## Writing contracts
+
+Methods, as well as attribute setters, can be associated with two types of contracts, entry contracts precondition (`expect`) and exit contracts postcondition (`ensure`). Together, they define the rules to be respected between the callee and the caller. If the caller fulfills the `expect` condition on entry then the caller can assume that the callee fulfill the `ensure` condition on exit.
+
+All contracts (`expect`, `ensure`) are evaluated in the same context as the called method, it's possible to refer to attributes, methods and parameters in the condition.
+
+When defining a contract (`expect`, `ensure`) on an attribute like bellow, only the automatically generated write property (setter) will use the contract.
+
+~~~
+	var baz: Int is expect(baz > 0) # In the contract, baz refers to the parameter of the write property (setter).
```

Comment:
If I understand properly I can use an expect and a ensure on a attribute, one will be used on the getter and the other one on the setter?

Reply:
When applying an expect or a ensure only the setter is impacted. It was not really clear which solution was best. It depends on the context, do we want to guarantee that the attribute has been correctly assigned or do we want to guarantee that the returned value is valid. In all cases it's possible to add contracts to the getter.

```nit
	redef fun attribute
	is
		ensure[...]
	do
		return _attribut
	end
```